packages/compiler/config/config-loader.ts

```diff
@@ -1,21 +1,30 @@
 import { getAnyExtensionFromPath, getBaseFileName, joinPaths } from "../core/path-utils.js";
 import { SchemaValidator } from "../core/schema-validator.js";
 import { CompilerHost, Diagnostic } from "../core/types.js";
-import { deepClone, deepFreeze, loadFile } from "../core/util.js";
+import { deepClone, deepFreeze, findProjectRoot, loadFile } from "../core/util.js";
 import { CadlConfigJsonSchema } from "./config-schema.js";
 import { CadlConfig } from "./types.js";
 
 const configFilenames = [".cadlrc.yaml", ".cadlrc.yml", ".cadlrc.json", "package.json"];
 const defaultConfig: CadlConfig = deepFreeze({
-  plugins: [],
   diagnostics: [],
   emitters: {},
-  lint: {
-    extends: [],
-    rules: {},
-  },
 });
 
+/**
+ * Load the cadl configuration for the provided directory
+ * @param host
+ * @param directoryPath
+ */
+export async function loadCadlConfigForPath(
+  host: CompilerHost,
+  directoryPath: string
+): Promise<CadlConfig> {
+  const projectRoot = await findProjectRoot(host, directoryPath);
+  const configDir = projectRoot ?? directoryPath;
+  return loadCadlConfigInDir(host, configDir);
```

Comment:
We should probably cache this like we cache the package.json.

To do so you will need also need to add to watched files to invalidate the cache, I think.
 https://github.com/microsoft/cadl/blob/5b23892fb6a82b67031272684e873660cab3e395/packages/cadl-vs/src/VSExtension.cs#L46
https://github.com/microsoft/cadl/blob/e3f808b899d46151b46cc5abff13a3104c659396/packages/cadl-vscode/src/extension.ts#L19


Do you still think we should change the main.cadl search to match or is it better to keep them separate and slightly different? I forget now what the issues were.

Reply:
so I don't think I actually updated the search for the main file. There was actually a reason it was done this way, I misunderstood the logic the 1st time.

What it does to look for the main is navigate up until it finds a package.json that tells where a valid main is or there is just a `main.cadl` file.

Without that logic our samples with multiple project wouldn't work. So we do need to keep that logic as far as I could tell.

Reply:
Sounds good. Along with the watched file changes, I was asking about caching the config object like we do for package.json object:

https://github.com/microsoft/cadl/blob/305e49946aca8c982d04994db59e0e3f43e550a3/packages/compiler/server/server.ts#L581-L588

But, now I think "extends" makes invalidation harder because the config can change without the root config file changing :( 

Anyway, I think we can optimize later. The real gains will come from caching parse results and we can think about this case. Consider filing something.

Reply:
We also aren't avoiding all the probing in directories, just the reading and parsing of JSON, so there's other room for improvement here. Probably makes sense to look at this all up along with https://github.com/Azure/cadl-azure/issues/619

Reply:
will add a note on that issue